### DISCUSSION TITLE
Support for FTDI based adapters

### DISCUSSION THREAD

**Discussion by electrodonkey** [pyocd/pyOCD#1337] [General]
Hello, i am a long time fan of pyOCD and have noticed that there were some various mentions about FTDI support.
Was there already some work done by the project team already? Is this planned or still open for contributions?
If still open, where should one start?

Thank you,
Dan

**Comment by flit** [comment]
Hi @electrodonkey. Apologies for the slow reply! I've been moving house (and recovering).

Support for FTDI based probes is on the wish list (and I even bought an FTDI cable to be able to test it). But no work has been done yet. Adding this would be fantastic!

First, we would need to define to scope. JTAG only or SWD too? I seem to remember SWD requires a small additional circuit? One issue is that full support for JTAG has never been implemented in pyocd—there just hasn't been a driving reason, even though it's been on the todo list. I'll be happy to help out on the JTAG support side if you can do the FTDI probe driver. (I have the very beginnings in my [feature/jtag](https://github.com/flit/pyOCD/tree/feature/jtag) branch.)

How to control the FTDI? [pyftdi](https://pypi.org/project/pyftdi/) is top of the list. However, the readme says Windows is not officially supported even though it works (probably due to USB driver issues, I bet you have to manually install a driver via Zadig for it to work). There are some other FTDI interface packages on PyPI, mostly wrappers around either libftdi or FTDI's official driver lib. Ultimately, the best option is probably to support multiple driver backends, similarly to how pyocd supports multiple backends for CMSIS-DAP.

All probe drivers are plugins that supply subclasses of `DebugProbe` in [`pyocd/probe/debug_probe.py`](https://github.com/pyocd/pyOCD/blob/main/pyocd/probe/debug_probe.py). The built-in probe driver plugins reside in the same package, but plugins from external packages are supported (pyocd uses setuptools plugin support). Check out `JLinkProbe` in [`jlink_probe.py`](https://github.com/pyocd/pyOCD/blob/main/pyocd/probe/jlink_probe.py) for a straightforward example. The plugin class `JLinkProbePlugin` is at the bottom of the file. You will naturally need to define a number of session options for configuring the FTDI pins and so on; these are defined in a list returned by the plugin class' `options` property.

If you'd like to talk about this more interactively, pyocd has a Slack workspace ([join](https://join.slack.com/t/pyocd/shared_invite/zt-zqjv6zr5-ZfGAXl_mFCGGmFlB_8riHA)). Regardless, I'll be more than happy to help with any questions!

**Reply by aseemjakhar** [reply]
Hi,

Is there any movement on this? Let me know if I can help.

Cheers...

**Reply by nattgris** [reply]
I just posted an announcement here https://github.com/pyocd/pyOCD/discussions/1701, feel free to pitch in.
